### ISSUE TITLE
Sponsorblocks applied to downloads

### ISSUE THREAD

**Issue by augmentedagent** [polymorphicshade/Tubular#117]
### Checklist

- [X] I made sure that there are *no existing issues* - [open](https://github.com/polymorphicshade/Tubular/issues) or [closed](https://github.com/polymorphicshade/Tubular/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have read the [FAQ](https://newpipe.net/FAQ/) and my problem isn't listed.
- [X] I'm aware that this is a request for Tubular itself and that requests for adding a new service need to be made in [vanilla NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor/issues).
- [X] I have taken the time to fill in all the required details. I understand that the feature request will be dismissed otherwise.
- [X] This issue contains only one feature request.
- [X] I have read and understood the vanilla NewPipe [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Feature description

Apply sponsorblocks to downloaded content. 

### Why do you want this feature?

Currently, when downloading content (specifically audio only), the content is downloaded and saved without applying sponsorblocks.

### Why ist the feature relevant to this fork?

This fork implements the sponsorblock functionality.

**Comment by HairyMilkshakes** [comment]
Is that even possible? Unless you're talking about playing downloaded videos through Tubular?

**Comment by augmentedagent** [comment]
No not proposing playing downloaded content through Tubular. I don't know how sponsorblock is implemented in Tubular, but it is able to be done through yt-dlp - the content is downloaded, then sponsorblock timecodes are applied and the resulting file is saved sans the blocked bits.  Knowing yt-dlp, this is probably done by running through ffmpeg with the cuts from sponsorblock, but orchestrated by yt-dlp. 

I was suggesting that Tubular might be able to do the same thing at the end of the download action.

**Comment by HairyMilkshakes** [comment]
Interesting, never knew that. That would be really good.

**Comment by qgustavor** [comment]
I should add one crucial information: that's really useful to download music without non-music segments. There were times I preferred downloading music over yt-dlp over Tubular only because it supports Sponsorblock.

**Comment by polymorphicshade** [comment]
In a version of NPxSB, I had an experimental feature of having an offline video player that worked with SponsorBlock.

It was a hack and didn't work very well... I have yet to implement something like this in Tubular because It think the core implementation should be in NewPipe (see: https://github.com/TeamNewPipe/NewPipe/issues/10542).

Nonetheless, I really like the idea of incorporating yt-dlp in Tubular. There are already have several offline video players on android, so just adding yt-dlp in Tubular would bridge the gap nicely I think.

**Comment by augmentedagent** [comment]
> There are several offline video players on android already, so just adding yt-dlp in Tubular would bridge the gap nicely I think.

Yes, and in my use case it's mainly for audio files for which there are also numerous players. Typically I just use the audio player within my file browser since it supports m4a nicely.

**Comment by canadiannomad** [comment]
I listen to podcasts that are downloaded..  This would be nice for that as well.

**Comment by nucspl** [comment]
If it's of any help, [yt-dlp](https://github.com/yt-dlp/yt-dlp) does have an in-built function that omits including SponsorBlock segments from the downloaded file itself. However, I don't think it's customisable or is able to read user-made segments. I think what could be done instead is download the pieces that aren't SponsorBlock segments — with respect to user-made segments via Tubular — using yt-dlp, then stitch them together via Ffmpeg. I don't know if Ffmpeg works on Android, though.

Seeing this issue is a little funny to me, actually. A few days ago, I was thinking of issuing a feature request to replace the NewPipe downloader with yt-dlp outright, and this issue was one of the things I cited in a draft — the others pertained to the ability to choose specific stream codecs.

**Comment by joelpeapen** [comment]
There's a good YouTube downloader app called [YTDLnis](https://github.com/deniscerri/ytdlnis) using yt-dlp that incorporates sponsor block

**Comment by rambalachandran** [comment]
I also recommend having a look at [seal](https://f-droid.org/en/packages/com.junkfood.seal/) for inspiration on integrating yt-dlp for Android

**Comment by qgustavor** [comment]
The guy from the last comment is a troll: just Google him, they posts a lot of comments on public repos blaming developers for issues without any consideration. No sane person wishing Sponsorblock would ever switch from Tubular with a partial implementation to NewPipe without anything of it implemented, thus I guessed it could be a troll. The Google search just confirmed my guess. They even made their GitHub activity private to hide their troll activity. Just ignore them.

**Comment by dftf-stu** [comment]
While this might be tricky to implement (given you can't just grab the file, it'd have to be encoded on-the-fly), it would be a cool feature to have, yes.

Some videos can literally have up to a third of the entire video being skipped, so would be great if we could not include any of that when downloading a video, to save bandwidth and storage-space.

Or, as an alternative:

Maybe downloaded videos could be playable within Tubular itself (not an external app) and the SponsorBlock skips could be saved into a local database and so still get applied when an offline, local video is then played?
